File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/optimize/generate.mjs
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { resolve, dirname } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ASSETS_DIR = resolve(__dirname, "..", "..", "assets");

let _bank = null;
function loadNegativeBank() {
  if (_bank) return _bank;
  const text = readFileSync(
    resolve(ASSETS_DIR, "negative_query_bank.txt"),
    "utf8",
  );
  _bank = text
    .split("\n")
    .map((s) => s.trim())
    .filter(Boolean);
  return _bank;
}

const VERB_RE =
  /\b(scaffold|validate|audit|score|emit|resolve|lint|optimize|export|generate|sort|detect|warn|run|build|parse|check|wire|install|configure|rewrite|propose|tune)\b/g;

const NOUN_RE =
  /\b(skill|frontmatter|description|dependency|dependencies|registry|prompt|properties|graph|cycle|inversion|tier|fixture|workflow|hook|gate|comment|provider|variant|hit-rate|score|optimizer|integrator)\b/g;

function unique(arr) {
  const seen = new Set();
  const out = [];
  for (const x of arr) {
    if (!seen.has(x)) {
      seen.add(x);
      out.push(x);
    }
  }
  return out;
}

export function extractVerbsAndNouns(parsed) {
  const body = (parsed.body ?? "").toLowerCase();
  const desc = (parsed.description ?? "").toLowerCase();
  const verbs = unique([
    ...(body.match(VERB_RE) ?? []),
    ...(desc.match(VERB_RE) ?? []),
  ]);
  const nouns = unique([
    ...(desc.match(NOUN_RE) ?? []),
    ...(body.match(NOUN_RE) ?? []),
  ]);
  return { verbs, nouns };
}

export function generateDeterministic(parsed, n = 8) {
  const { verbs, nouns } = extractVerbsAndNouns(parsed);
  const positive = [];
  for (let i = 0; i < n; i++) {
    const v = verbs[i % Math.max(1, verbs.length)] ?? "do";
    const noun = nouns[i % Math.max(1, nouns.length)] ?? "thing";
    positive.push(`${v} the ${noun}`);
  }
  const bank = loadNegativeBank();
  const negative = bank.slice(0, n);
  return { positive, negative };
}

File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/optimize/generate.py
from __future__ import annotations

import re
from pathlib import Path

_ASSETS_DIR = Path(__file__).resolve().parent.parent / "assets"

_VERB_RE = re.compile(
    r"\b(scaffold|validate|audit|score|emit|resolve|lint|optimize|export|generate|sort|detect|warn|run|build|parse|check|wire|install|configure|rewrite|propose|tune)\b"
)
_NOUN_RE = re.compile(
    r"\b(skill|frontmatter|description|dependency|dependencies|registry|prompt|properties|graph|cycle|inversion|tier|fixture|workflow|hook|gate|comment|provider|variant|hit-rate|score|optimizer|integrator)\b"
)

_bank = None


def _load_negative_bank() -> list[str]:
    global _bank
    if _bank is not None:
        return _bank
    text = (_ASSETS_DIR / "negative_query_bank.txt").read_text(encoding="utf-8")
    _bank = [s.strip() for s in text.split("\n") if s.strip()]
    return _bank


def _unique(arr):
    seen = set()
    out = []
    for x in arr:
        if x not in seen:
            seen.add(x)
            out.append(x)
    return out


def extract_verbs_and_nouns(parsed) -> tuple[list[str], list[str]]:
    body = (parsed.body or "").lower()
    desc = (parsed.description or "").lower()
    verbs = _unique(_VERB_RE.findall(body) + _VERB_RE.findall(desc))
    nouns = _unique(_NOUN_RE.findall(desc) + _NOUN_RE.findall(body))
    return verbs, nouns


def generate_deterministic(parsed, n: int = 8) -> dict:
    verbs, nouns = extract_verbs_and_nouns(parsed)
    positive: list[str] = []
    for i in range(n):
        v = verbs[i % max(1, len(verbs))] if verbs else "do"
        noun = nouns[i % max(1, len(nouns))] if nouns else "thing"
        positive.append(f"{v} the {noun}")
    bank = _load_negative_bank()
    negative = bank[:n]
    return {"positive": positive, "negative": negative}
